pennylane/devices/capabilities.py

```diff
@@ -0,0 +1,328 @@
+    observables: dict[str, OperatorProperties] = field(default_factory=dict)
+    measurement_processes: dict[str, list[ExecutionCondition]] = field(default_factory=dict)
+    qjit_compatible: bool = False
+    mid_circuit_measurements: bool = False
+    runtime_code_generation: bool = False
+    dynamic_qubit_management: bool = False
+    overlapping_observables: bool = True
+    non_commuting_observables: bool = False
+    initial_state_prep: bool = False
+    options: dict[str, any] = field(default_factory=dict)
+
+    def filter(self, finite_shots: bool) -> "DeviceCapabilities":
+        """Returns the device capabilities conditioned on the given program features."""
+
+        return (
+            self._exclude_entries_with_condition(ExecutionCondition.ANALYTIC_MODE_ONLY)
+            if finite_shots
+            else self._exclude_entries_with_condition(ExecutionCondition.FINITE_SHOTS_ONLY)
+        )
+
+    def _exclude_entries_with_condition(
+        self, condition: ExecutionCondition
+    ) -> "DeviceCapabilities":
+        """Removes entries from the capabilities that has the given condition."""
+
+        operations = {k: v for k, v in self.operations.items() if condition not in v.conditions}
+        observables = {k: v for k, v in self.observables.items() if condition not in v.conditions}
+        measurement_processes = {
+            k: v for k, v in self.measurement_processes.items() if condition not in v
+        }
+        return replace(
+            self,
+            operations=operations,
+            observables=observables,
+            measurement_processes=measurement_processes,
+        )
+
+    @classmethod
+    def from_toml_file(cls, file_path: str, runtime_interface="pennylane") -> "DeviceCapabilities":
+        """Loads a DeviceCapabilities object from a TOML file.
+
+        Args:
+            file_path (str): The path to the TOML file.
+            runtime_interface (str): The runtime execution interface to get the capabilities for.
+                Acceptable values are "pennylane" and "qjit". Use "pennylane" for capabilities of
+                the device's implementation of `Device.execute`, and "qjit" for capabilities of
+                the runtime execution function used by a qjit-compiled workflow.
+
+        """
+        document = load_toml_file(file_path)
+        capabilities = parse_toml_document(document)
+        update_device_capabilities(capabilities, document, runtime_interface)
+        return capabilities
+
+
+VALID_COMPILATION_FLAGS = {
+    "qjit_compatible": False,
+    "mid_circuit_measurements": False,
+    "runtime_code_generation": False,
+    "dynamic_qubit_management": False,
+    "overlapping_observables": True,
+    "non_commuting_observables": False,
+    "initial_state_prep": False,
+}
+
+
+def _get_toml_section(document: dict, path: str, prefix: str = "") -> dict:
+    """Retrieves a section from a TOML document using a given path."""
```

Comment:
I guess "path" is official terminology for the TOML format? It might be nice to clarify this a little bit. The documentation feels a little sparse.
 
```suggestion
    """Retrieves a section from a TOML document using a given path (such as "operator.gates"), and returns it as a dictionary. 
    
    If a prefix is provided, it will be included at a the beginning of the path; for example, for `path="operator.gates"` and `str="qjit"`, the `[qjit.operator.gates]` section of the TOML file will be returned as a dictionary"""
```